frontend/rolecall/src/app/user/user-editor.component.ts

```diff
@@ -1,121 +1,197 @@
-import { ActivatedRoute } from '@angular/router';
-import { isNullOrUndefined } from 'util';
-import { User, UserApi } from '../api/user_api.service';
-
+import {Location} from '@angular/common';
+import {Component, EventEmitter, OnInit} from '@angular/core';
+import {ActivatedRoute} from '@angular/router';
+import {User, UserApi} from '../api/user_api.service';
 
 /**
  * The view for the User Editor, allowing users to create other users
- * and view user information
+ * and view user information.
  */
 @Component({
   selector: 'app-user-editor',
   templateUrl: './user-editor.component.html',
   styleUrls: ['./user-editor.component.scss']
 })
 export class UserEditor implements OnInit {
-
   permissionsSet: EventEmitter<string[]> = new EventEmitter();
   rolesSet: EventEmitter<string[]> = new EventEmitter();
 
   currentSelectedUser: User;
   renderingUsers: User[];
-  urlPointingUUID: string;
+  private urlPointingUUID: string;
+
+  private prevWorkingState: User | undefined;
+  private workingUser: User | undefined;
+  disableSave = true;
+  creatingUser = false;
+
+  usersLoaded = false;
+  private dataLoaded = false;
 
-  prevWorkingState: User;
-  workingUser: User;
-  disableSave: boolean = true;
-  creatingUser: boolean = false;
+  private lastSelectedUserEmail: string;
 
-  usersLoaded: boolean = false;
-  dataLoaded: boolean = false;
+  private nameToPropertyMap = {
+    'First Name': {
+      key: 'first_name',
+      type: 'string'
+    },
+    'Last Name': {
+      key: 'last_name',
+      type: 'string'
+    },
+    Email: {
+      key: 'contact_info.email',
+      type: 'string'
+    },
+    'Date Joined': {
+      key: 'date_joined',
+      type: 'date'
+    },
+    Phone: {
+      key: 'contact_info.phone_number',
+      type: 'string'
+    },
+    Roles: {
+      key: 'has_roles',
+      type: 'roles'
+    },
+    Permissions: {
+      key: 'has_permissions',
+      type: 'permissions'
+    },
+  };
+
+  rolesNamesMap = {
+    isAdmin: 'Is Admin',
+    isCoreographer: 'Is Coreographer',
+    isDancer: 'Is Dancer',
+    isOther: 'Is Other',
+  };
+  permissionsNamesMap = {
+    canLogin: 'Can Login',
+    canReceiveNotifications: 'Receives Notifications',
+    managePerformances: 'Manages Performances',
+    manageCasts: 'Manages Casts',
+    managePieces: 'Manages Pieces',
+    manageRoles: 'Manages Roles',
+    manageRules: 'Manages Rules',
+  };
 
-  lastSelectedUserEmail: string;
+  private currentDate = Date.now();
 
-  constructor(private route: ActivatedRoute, private userAPI: UserApi,
-    private location: Location) { }
+  constructor(
+      private readonly route: ActivatedRoute,
+      private readonly userAPI: UserApi,
+      private readonly location: Location) {
+  }
 
   ngOnInit(): void {
-    let uuid = this.route.snapshot.params.uuid;
-    if (!isNullOrUndefined(uuid)) {
+    const uuid: string = this.route.snapshot.params.uuid;
+    if (uuid) {
```

Comment:
You don't always specify a type when assigning a const. Why here?

Reply:
Good eye. The reason for changing it here is that `params` has the type `"[key: string]: any"` so the value of `uuid` evaluates to `"any"`. Narrowing the type as early as possible allows the rest of the code to know that it's actually a string for proper type checking.